### PR TITLE
BUG: resampling with NaT in TimedeltaIndex (#13223)

### LINKED ISSUE #13223: Resampling with NaT in TimedeltaIndex raises MemoryError
`DatetimeIndex` can handle `NaT`:

```
In [2]: dti = pd.DatetimeIndex(['1970-01-01 00:00:00', pd.NaT, '1970-01-01 00:00:02'])

In [3]: pd.DataFrame([2, 3, 5], dti).resample('1s').mean()
Out[3]: 
                       0
1970-01-01 00:00:00  2.0
1970-01-01 00:00:01  NaN
1970-01-01 00:00:02  5.0
```

But `TimedeltaIndex` can't:

```
In [4]: tdi = pd.to_timedelta(['0s', pd.NaT, '2s'])

In [5]: pd.DataFrame([2, 3, 5], tdi).resample('1s').mean()
...
MemoryError:
```

Versions:

```
python: 2.7.11.final.0
OS: Linux
machine: x86_64
pandas: 0.18.1+38.gc7e6f5f.dirty
numpy: 1.10.2
```

### PR DESCRIPTION
**Pull request by fr-chaofan**
- [ ] closes #13223 
 - [ ] tests added / passed
 - [ ] passes ``git diff upstream/master | flake8 --diff``
 - [ ] whatsnew entry

### CODE REVIEW

File: pandas/tseries/resample.py [hunk 1]
```diff
@@ -1204,6 +1206,13 @@ def _get_time_delta_bins(self, ax):
         end_stamps = labels + 1
         bins = ax.searchsorted(end_stamps, side='left')
 
+        if ax.hasnans:
+            binner = binner.insert(0, tslib.NaT)
+            labels = labels.insert(0, tslib.NaT)
+
+            n_NaT = sum([ax_i is tslib.NaT for ax_i in ax])
```
Comment (jreback) [inline pandas/tseries/resample.py]: ``n_NaT = ax.isnull().sum()``

File: pandas/tseries/resample.py [hunk 2]
```diff
@@ -1204,6 +1206,13 @@ def _get_time_delta_bins(self, ax):
         end_stamps = labels + 1
         bins = ax.searchsorted(end_stamps, side='left')
 
+        if ax.hasnans:
```
Comment (jreback) [inline pandas/tseries/resample.py]: though we actually need to ignore the nans, doesn't this actually create a NaT group? this is inconcistent with grouping.

File: pandas/tseries/tests/test_resample.py [hunk 3]
```diff
@@ -970,6 +970,15 @@ def test_resample_timedelta_idempotency(self):
         expected = series
         assert_series_equal(result, expected)
 
+    def test_resample_timedelta_missing_values(self):
+        # GH 13223
+        index = pd.to_timedelta(['0s', pd.NaT, '2s'])
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: test with an all NaT index

File: pandas/tseries/resample.py (hunk 1 above)
Comment (liuguangyu0220) [inline pandas/tseries/resample.py]: Will change. Thanks

File: pandas/tseries/resample.py (hunk 2 above)
Comment (liuguangyu0220) [inline pandas/tseries/resample.py]: NaT group will be ignored in aggregation. It is handled the same way as in function `_get_time_bins`.

File: pandas/tseries/resample.py [hunk 4]
```diff
@@ -1189,13 +1190,18 @@ def _get_time_delta_bins(self, ax):
             raise TypeError('axis must be a TimedeltaIndex, but got '
                             'an instance of %r' % type(ax).__name__)
 
+        if len(ax) > 0 and all(ax._isnan):
```
Comment (jreback) [inline pandas/tseries/resample.py]: I would give a better error message here. Is this consistent with how we handle all-nan grouping?

Comment (liuguangyu0220) [inline pandas/tseries/resample.py]: All-nan grouping doesn't seem to be handled elsewhere. Any suggestions?

Comment (jreback) [inline pandas/tseries/resample.py]: we currently ignore all nan groups, that's why I think it is there, so this is consistent, maybe a comment is worth it here.

Comment (jreback) [inline pandas/tseries/resample.py]: so this is *all* data is nan. Hmm, I would just give a better error message then.

File: pandas/tseries/tests/test_resample.py [hunk 5]
```diff
@@ -1007,6 +1007,20 @@ def test_resample_timedelta_idempotency(self):
...
+        series = pd.Series([2, 3, 5], index=index)
+        result = series.resample('1s').mean()
+        expected = pd.Series([2, np.nan, 5], index=pd.timedelta_range(
+            start='0s', end='2s', freq='1s'))
+        assert_series_equal(result, expected)
+
+        # all NaT
+        index = pd.to_timedelta([pd.NaT, pd.NaT, pd.NaT])
+        series = pd.Series([2, 3, 5], index=index)
+        self.assertRaises(DataError, series.resample('1s').mean)
```
Comment (jreback) [inline pandas/tseries/tests/test_resample.py]: This is probably the right idea (e.g. raising a nice message),

e.g. we do this for all-nan groups (which is unfriendly)

```
In [8]: s = Series([1,2,3],[np.nan,np.nan,np.nan])

In [9]: s.groupby(s.index).sum()
TypeError: unhashable type: 'Float64Index'
```

**Comment by liuguangyu0220** [comment]
Rebased

File: pandas/tseries/resample.py (hunk 4 above)
Comment (jreback) [inline pandas/tseries/resample.py]: use ``ax.hasnans``

File: pandas/tseries/resample.py (hunk 2 above)
Comment (jreback) [inline pandas/tseries/resample.py]: add a comment here on what you are doing

File: pandas/tseries/resample.py [hunk 6]
```diff
@@ -1219,13 +1220,18 @@ def _get_time_delta_bins(self, ax):
...
+            raise DataError('all-nan groupings not valid')
+
         if not len(ax):
             binner = labels = TimedeltaIndex(
                 data=[], freq=self.freq, name=ax.name)
             return binner, [], labels
 
-        start = ax[0]
-        end = ax[-1]
+        # Addresses GH #13223
```
Comment (jreback) [inline pandas/tseries/resample.py]: add a comment on what you are doing here (and why not just selecting start/end)

**Review by jreback — CHANGES_REQUESTED** [review]
looks pretty good. some doc fixups

pls add a whatsnew entry in bug fixes (0.20.0).

ping on green.

**Comment by jreback** [comment]
can you rebase / upate

**Comment by liuguangyu0220** [comment]
rebased

**Comment by jreback** [comment]
when you rebase you should end up with a small number of commits.

try

```
git rebase -i yourbranchname origin/master
git push yourremote yourbranchname
```

**Comment by jreback** [comment]
closing. this needs a new PR with a cherry-pick on top of master.

**Pull request closed without merging**